### ISSUE TITLE
Unable to use kernel installed in debian-8 VM

### ISSUE THREAD

**Issue by ctrlaltdel** [QubesOS/qubes-issues#3178]
#### Qubes OS version:
R4.0

#### Affected TemplateVMs: 
debian-8

---

### Steps to reproduce the behavior:

```
$ qvm-prefs debian-8 kernel ''
$ qvm-prefs debian-8 debug true
$ qvm-run debian-8 gnome-terminal
```
### Expected behavior:

VM boots correctly with the guest provided kernel which makes it easy to use out-of-tree kernel modules (like wireguard) inside VMs.

### Actual behavior:

VM boot fails with the following error message:

```
/scripts/local-top/qubes_cow_setup: line 106: can't create /etc/udev/rules.d/99-root.rules: nonexistent directory
Gave up waiting for root device.  Common problems:
[...]
```

Digging from the initramfs shell shows that `/dev/mapper/dmroot` doesn't exist, don't know how it's supposed to be created?

### General notes:

I've followed [documentation](https://www.qubes-os.org/doc/managing-vm-kernel/#using-kernel-installed-in-the-vm) on how to use kernel installed in the VM.

### Work-around

Manually changing `/dev/mapper/dmroot` to `/dev/xvda1` in `/etc/default/grub.d/30-qubes.cfg` makes the VM boot correctly. Not sure about the side-effects of this though.

**Comment by marmarek** [comment]
> Manually changing /dev/mapper/dmroot to /dev/xvda1 in /etc/default/grub.d/30-qubes.cfg makes the VM boot correctly. Not sure about the side-effects of this though.

AppVMs based on such templates (and having `kernel` set to '') will be broken.

**Comment by awokd** [comment]
@andrewdavidwong Not sure if this is a regression or a separate issue, but I'm working on updating https://www.qubes-os.org/doc/managing-vm-kernel/. Fedora-26 worked as expected, but debian-9 gives me an error when trying to install a different kernel:
~~~
/etc/kernel/postinst.d/zz-update-grub:
/usr/sbin/grub-probe: error: failed to get canonical path of `/dev/mapper/dmroot'.
run-parts: /etc/kernel/postinst.d/zz-update-grub exited with return code 1
dpkg: error processing package linux-image-4.14.0-0.bpo.3-amd64 (--configure):
~~~
This step used to result in a non-fatal error of:
~~~
grub2-probe: error: cannot find a GRUB drive for /dev/mapper/dmroot. Check your device.map
~~~
But now it seems to be fatal. I see there is a test for `systemd-detect-virt --quiet --container` in `/etc/kernel/postinst.d/zz-update-grub` which appears to return `none` under a 4.0 HVM. Was this different in a 3.2 HVM, or is it something entirely unrelated?

**Comment by andrewdavidwong** [comment]
@awokd: Good question. I don't know. Reopening to see what @marmarek thinks.

**Comment by awokd** [comment]
@tasket Any suggestions on my comment above? It's one of the last "4.0" documents on my list.

**Comment by redshiftzero** [comment]
For SecureDrop's [Qubes journalist workstation](https://github.com/freedomofpress/securedrop-workstation) we'd like to use a custom hardened kernel for Debian templates - any update on the status of this ticket?

**Comment by marmarek** [comment]
Sadly no progress here. I'll look into this next week (ping me on Wed if I don't update here).

**Comment by marmarek** [comment]
A workaround: before executing `update-grub2`, make a symlink: `ln -s ../xvda3 /dev/mapper/dmroot`.
For me the symlink is in place when booting dom0-provided kernel, but it isn't anymore when booting debian kernel (and initramfs generated by debian). So, first setup works, but then kernel update fails.

The symlink is correct inside TemplateVM/StandaloneVM only (but not in TemplateBasedVM). But that's ok because it doesn't make sense to update grub there as changes in /boot wouldn't survive VM restart.

**Comment by marmarek** [comment]
@redshiftzero can you confirm the symlink helps in your case?

**Comment by redshiftzero** [comment]
Excellent, thank you @marmarek, we'll try this out and report back (cc @emkll)

**Comment by awokd** [comment]
FWIW, I'm trying to test this out as well but after cloning the debian-9 template (at current stable patches) and the:
~~~
qvm-prefs <clonetemplatename> virt_mode hvm
qvm-prefs <clonetemplatename> kernel ''
~~~
steps, I can no longer get a usable terminal window to come up. It just has diagonal lines of gibberish. I enabled debug mode and set qrexec timeout to 120. It seems to be trying to create the swap space on every boot now, instead of just on first boot. I can see in the debug window it `Gave up waiting on` something, but the message goes by too fast to catch the rest and the logs aren't helpful.

**Comment by marmarek** [comment]
> steps, I can no longer get a usable terminal window to come up. It just has diagonal lines of gibberish.

Make sure to build u2mfn module for the kernel you want to use, in some cases dkms fails to do that automatically. `dkms autoinstall -k <kernel version>`

**Comment by awokd** [comment]
I switched the clone back to PVH, ran `dkms autoinstall -k 4.9.0-6-amd64`, then set back to HVM. I finally caught the error message on boot. After a ~30 second pause in the boot sequence, it said `Gave up waiting for suspend/resume device`, but I then had a usable terminal window. Next,
~~~
apt install linux-headers-rt-amd64
apt install linux-image-rt-amd64
~~~
worked! I have a terminal window running `4.9.0-6-rt-amd64 #1 SMP PREEMPT RT Debian 4.9.88-1+deb9u1 (2018-05-07) x86_64 GNU/Linux`.

This is a stock debian-9 template with just the standard Qubes and Debian updates. I've noticed DKMS errors a couple times when `apt upgrade`ing the kernel on it. I can document around that, but is it something related to https://github.com/QubesOS/qubes-issues/issues/2844?

That ~30 second hang on boot isn't going away either. On R3.2 it was just the first boot that took a long time; subsequent ones were fast. A random [thread](https://www.reddit.com/r/debian/comments/64wonb/gave_up_waiting_for_suspendresume_device/) suggests it's caused when the swap device isn't set to the correct UUID. Tried [these steps](https://www.qubes-os.org/doc/managing-vm-kernel/#installing-kernel-in-debian-vm-r40-work-in-progress) but didn't help. I'll attempt to keep troubleshooting this unless you/anyone have a better idea.

**Comment by marmarek** [comment]
> I can document around that, but is it something related to #2844?

It looks like it. Note that dom0-provided kernel already has u2mfn module, so this error isn't a problem on its own. But if it prevents building the module for other kernels, it is a problem.

> It seems to be trying to create the swap space on every boot now, instead of just on first boot. I can see in the debug window it Gave up waiting on something

Yes, now VM get empty volatile image and it is initialized in initramfs (instead of using pre-generated Fedora specific one). I see two options:
 - move [volatile volume initialization](https://github.com/QubesOS/qubes-linux-utils/blob/master/dracut/full-dmroot/qubes_cow_setup.sh#L72-L113) - including mkswap call somewhere earlier (it's linked [as local-top](https://github.com/QubesOS/qubes-linux-utils/tree/master/initramfs-tools/local-top)
 - convince Debian's initramfs to not wait for resume device (`noresume` kernel option?)

**Comment by awokd** [comment]
Would `noresume` on these VMs affect Qubes/system-wide hibernation, or vice-versa? If not, that might be the simplest approach. I don't use hibernate myself, so can test if that fixes this particular issue but not system-wide.

**Comment by marmarek** [comment]
No, VMs don't do suspend to disk using own swap, in none of Qubes version/features combination.

**Comment by awokd** [comment]
Adding `noresume` to grub.cfg boot option resolved that 30 second hang. I can work on a code PR for this approach, if you like, or add to the documentation instead?

**Comment by marmarek** [comment]
We already have place for this: https://github.com/QubesOS/qubes-core-agent-linux/blob/master/misc/grub.qubes

**Comment by emkll** [comment]
@marmarek : The symlink workaround you describe does address the issue. Thanks!

**Comment by awokd** [comment]
Updated a clone of debian-9 to current-testing and symlink works- no more fatal errors on a grub update. I don't see `noresume` in `/etc/default/grub` or `grub.d/30-qubes.cfg`, though. Would I need to rebuild and reinstall the template?

**Comment by marmarek** [comment]
That one package (qubes-core-agent) wasn't part of that update. Should be in qubes-core-agent-4.0.32
